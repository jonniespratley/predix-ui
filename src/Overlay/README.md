The `<Overlay/>` is a component used for modal dialogues, popovers, and other use cases where the remainder of the application content is obscured or falls away from the primary actionable component. The overlay covers the entire screen.


## Usage

```react
state:
  open: false
---
const {Button, Overlay} = PxReact;
<div>
  <Button
    onClick={(e) => setState({open: !state.open})}>Open Overlay</Button>
  <Overlay
    opened={state.open}
    onOverlayClick={(e) => setState({open: !state.open})}/>
</div>
```


## Properties

```table
span: 6
rows:
  - Name: opened
    Type: bool
    Description: The visibility of the modal.

  - Name: ignoreEscapeKeyUp
    Type: bool
    Description: If true, ESC key will not close modal.

  - Name: ignoreOverlayClick
    Type: bool
    Description: If true, clicking backdrop will not close modal.

  - Name: onRequestClose
    Type: function
    Description: Invoked when request to close overlay happens.

  - Name: onEscapeKeyUp
    Type: function
    Description: Invoked when ESC key is up.

```
