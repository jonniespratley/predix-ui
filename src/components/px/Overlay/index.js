import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

const Overlay = styled.div`
  position        : fixed;
  top             : 0;
  right           : 0;
  bottom          : 0;
  left            : 0;
  z-index         : 900;
  background-color: var(--px-modal-overlay-color, rgba(0, 0, 0, 0.5));
  transition      : background-color 0.2s cubic-bezier(.78,.13,.16,.87);
  ${props => props.invisible && css`
     visibility      : hidden;
     background-color: transparent;
  `}
`;
Overlay.defaultProps = {
  className: 'px-overlay'
};
Overlay.displayName = 'Overlay';

class OverlayComponent extends React.Component {
  constructor(props) {
    super(props, { displayName: 'Overlay' });
    this.state = {
      opened: props.opened
    };
    this._handleOverlayClick = this._handleOverlayClick.bind(this);
    this._handleEscKeyUp = this._handleEscKeyUp.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    this.setState({ opened: nextProps.opened });
  }

  _handleEscKeyUp(e) {
    if (e.keyCode === 27 && !this.props.ignoreEscapeKeyUp) {
      if (this.props.onEscapeKeyUp) {
        this.props.onEscapeKeyUp(e);
      }
    }
  }

  _handleOverlayClick(e) {
    if (this.props.onOverlayClick && !this.props.ignoreOverlayClick) {
      this.props.onOverlayClick(e);
    }
  }

  close() {
    if (this.props.onRequestClose) {
      this.props.onRequestClose(this.state);
    }
    this.setState({ opened: false });
  }

  open() {
    this.setState({ opened: true });
  }

  render() {
    const {
      children,
      fullScreen
    } = this.props;
    const {
      opened
    } = this.state;
    return (
      <Overlay
        onKeyUp={this._handleEscKeyUp}
        fullScreen={fullScreen}
        invisible={!opened}
        onClick={this._handleOverlayClick}
      >
        {children}
      </Overlay>
    );
  }
}

OverlayComponent.defaultProps = {
  children: null,
  opened: false,
  fullScreen: false,
  onEscapeKeyUp: null,
  onOverlayClick: null,
  onRequestClose: null,
  ignoreEscapeKeyUp: false,
  ignoreOverlayClick: false
};

OverlayComponent.propTypes = {
  children: PropTypes.node,
  opened: PropTypes.bool,
  fullScreen: PropTypes.bool,
  onEscapeKeyUp: PropTypes.func,
  onOverlayClick: PropTypes.func,
  onRequestClose: PropTypes.func,
  ignoreEscapeKeyUp: PropTypes.bool,
  ignoreOverlayClick: PropTypes.bool
};

OverlayComponent.displayName = 'Overlay';

export default OverlayComponent;
