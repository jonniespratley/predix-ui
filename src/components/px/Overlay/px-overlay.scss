/**
 * px-overlay styles
 */
 .px-overlay {
   position        : fixed;
   top             : 0;
   right           : 0;
   bottom          : 0;
   left            : 0;
   z-index         : 900;
   background-color: var(--px-modal-overlay-color, rgba(0, 0, 0, 0.5));
   transition      : background-color 0.2s cubic-bezier(.78,.13,.16,.87);
   &--invisible{
     visibility      : hidden;
     background-color: transparent;
   }
 }
