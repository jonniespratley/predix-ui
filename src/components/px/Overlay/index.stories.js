import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, text, boolean, number } from '@storybook/addon-knobs/react';
import Overlay from './';

import { action } from '@storybook/addon-actions';
import { withInfo } from '@storybook/addon-info';

storiesOf('Overlay', module)
	.addDecorator(withKnobs)
	.add('default', () => (
		<Overlay 
			opened={boolean('opened', false)}
			onOverlayClick={action('clicked')}/>
	));
