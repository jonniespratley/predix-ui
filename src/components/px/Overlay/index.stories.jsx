import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean } from '@storybook/addon-knobs/react';
import { action } from '@storybook/addon-actions';
import { withReadme } from 'storybook-readme';

import README from './README.md';
import Overlay from './';

storiesOf('Components / Overlay', module)
  .addDecorator(withKnobs)
  .addDecorator(withReadme(README))
  .addWithJSX('default', () => (
    <Overlay
      opened={boolean('opened', false)}
      onOverlayClick={action('clicked')}
    />
  ))
  .addWithJSX('with content', () => (
    <Overlay
      opened={boolean('opened', false)}
      onOverlayClick={action('clicked')}
    >
      <p style={{
        textAlign: 'center',
        backgroundColor: 'white',
        padding: 20,
        margin: 20
      }}
      >
        This can be content
      </p>
    </Overlay>
  ));
