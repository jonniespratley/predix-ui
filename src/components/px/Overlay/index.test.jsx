import renderer from 'react-test-renderer';
import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';

import Overlay from './';

describe('Overlay', () => {

  test('matches snapshot', () => {
    const tree = renderer.create(
      <Overlay opened/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should has open class', () =>{
    const wrapper = shallow(<Overlay opened/>);
    expect(wrapper.state('opened')).toEqual(true);
  });

  test('onOverlayClick triggers on backdrop click', () => {
    const onButtonClick = sinon.spy();
    const onRequestClose = sinon.spy();
    const wrapper = shallow(<Overlay opened onRequestClose={onRequestClose} onOverlayClick={onButtonClick}/>);
    wrapper.simulate('click');

    expect(onButtonClick.calledOnce).toEqual(true);

    wrapper.instance().close();
    expect(onRequestClose.calledOnce).toEqual(true);
    expect(wrapper.state().opened).toEqual(false);

    wrapper.instance().open();
    expect(wrapper.state().opened).toEqual(true);
  });

  test('onEscapeKeyUp - fires event when pressed', () => {
    const onEscapeKeyUp = sinon.spy();
    const wrapper = shallow(<Overlay opened onEscapeKeyUp={onEscapeKeyUp} />);
    wrapper.simulate('keyUp', {key: 'ESC', keyCode: 27});
    expect(onEscapeKeyUp.calledOnce).toEqual(true);
  });

  test('does not close on ignoreBackdropClick click', () => {
    const onButtonClick = sinon.spy();
    const wrapper = shallow(<Overlay opened ignoreOverlayClick onOverlayClick={onButtonClick}/>);
    wrapper.simulate('click');
    expect(onButtonClick.calledOnce).toEqual(false);
  });

});
