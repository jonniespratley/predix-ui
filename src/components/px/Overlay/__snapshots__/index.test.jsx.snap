// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay matches snapshot 1`] = `
.c0 {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 900;
  background-color: var(--px-modal-overlay-color,rgba(0,0,0,0.5));
  -webkit-transition: background-color 0.2s cubic-bezier(.78,.13,.16,.87);
  transition: background-color 0.2s cubic-bezier(.78,.13,.16,.87);
}

<div
  className="px-overlay c0"
  onClick={[Function]}
  onKeyUp={[Function]}
/>
`;
